File: frontend/src/services/events.ts
// frontend/src/services/events.ts
import { apiFetch } from "@/lib/api";

/* ========= TYPES ========= */

/**
 * MUST MATCH backend/app/api/schemas/events.py -> SessionCreate
 */
export interface SessionCreatePayload {
  session_topic?: string | null;
  session_speaker_id?: number | null;

  // Date + time split (matches DB + backend schema)
  session_date: string; // YYYY-MM-DD
  session_start_time: string; // HH:mm:ss
  session_end_time: string; // HH:mm:ss

  // Location FKs (required by DB constraints)
  session_building_id: number;
  session_floor_id: number;
  session_room_id: number;
}

/**
 * MUST MATCH backend/app/api/schemas/events.py -> EventCreate
 */
export interface EventCreatePayload {
  event_name: string;
  event_description?: string;
  start_datetime: string;
  end_datetime: string;
  sessions?: SessionCreatePayload[];
}
export interface EventSessionDetails {
  id: number;
  topic: string;
  date: string;
  start_time: string;
  end_time: string;
  speaker?: {
    id: number;
    name: string;
    email?: string;
  } | null;
  venue?: {
    building?: string;
    floor?: string;
    room?: string;
  } | null;
}

export interface EventDetails {
  id: number;
  event_name: string;
  event_description?: string;
  start_datetime: string;
  end_datetime: string;
  status: string;
  sessions: EventSessionDetails[];
}

/* ========= API CALLS ========= */

export const getEvents = async () => {
  return await apiFetch("/api/v1/events");
};

export const getEventDetails = (eventId: number) => {
  return apiFetch(`/api/v1/events/${eventId}`);
};

export const createEvent = (payload: EventCreatePayload) => {
  return apiFetch("/api/v1/events", {
    method: "POST",
    body: JSON.stringify(payload),
  });
};

export const publishEvent = (eventId: number) => {
  return apiFetch(`/api/v1/events/${eventId}/publish`, {
    method: "PATCH",
  });
};

File: backend/app/api/schemas/events.py
# backend/app/api/schemas/events.py
from pydantic import BaseModel
from datetime import datetime
from typing import List, Optional, Literal


from datetime import date, time

class SessionCreate(BaseModel):
    session_topic: Optional[str]
    session_speaker_id: Optional[int]
    session_date: date
    session_start_time: time
    session_end_time: time
    session_building_id: int
    session_floor_id: int
    session_room_id: int



class EventCreate(BaseModel):
    event_name: str
    event_description: Optional[str]
    start_datetime: datetime
    end_datetime: datetime
    status: Optional[Literal["draft", "upcoming", "ongoing"]] = "draft"
    sessions: Optional[List[SessionCreate]]



class SessionOut(SessionCreate):
    id: int


class EventOut(BaseModel):
    id: int
    event_name: str
    event_description: Optional[str]
    start_datetime: datetime
    end_datetime: datetime
    place_id: Optional[int]
    status: str
    sessions: List[SessionOut]
